practical/ProcessMining/group2/conformance_checking/src/comparison.py

```diff
@@ -0,0 +1,234 @@
+from generate_footprint import FootPrintMatrix
+from check_conformance import ConformanceChecking
+from visualize_matrix import visualize_sorted_dict
+from pm4py.visualization.petri_net import visualizer
+from pm4py.objects.log.util import split_train_test
+from itertools import combinations
+from pm4py.algo.simulation.playout.petri_net import variants
+
+
+class Comparison:
+
+    def pipeline(self, input_log, algorithm: AlgoPm4Py, variant=variants.extensive):
+        """
+        Execute a process mining pipeline on an input log using a specified algorithm.
+
+        This method performs the following steps:
+        1. Generate an original FootPrintMatrix from the input log.
+        2. Run the specified algorithm on the input log to obtain a process model.
+        3. Replay the resulting model to obtain replayed logs.
+        4. Generate a FootPrintMatrix from the replayed logs.
+
+        Parameters:
+        ----------
+        input_log : EventLog
+            The input log in EventLog form (pm4py).
+        algorithm : AlgoPm4Py
+            The process mining algorithm to be applied to the input log.
+        variant : variants (optional)
+            The variant of the replay algorithm to be used. Default is variants.extensive.
+
+        Returns:
+        -------
+        tuple
+            A tuple containing the original FootPrintMatrix and the replayed FootPrintMatrix.
+        """
+        # Pipeline:
+        # 1. Get footprint matrix from log
+        fpm_original = FootPrintMatrix(input_log)
+        fpm_original.generate_footprint()
+        # 2. Run algorithm on log
+        net, start, end = get_model_from_pm4py(input_log, algorithm)
+        # 3. Replay resulting model from algorithm
+        logs_replayed = get_traces_with_replay(net, start, end, variant)
+        # 4. Get footprint matrix from replayed log
+        fpm_replayed = FootPrintMatrix(logs_replayed)
+        fpm_replayed.generate_footprint()
+        return fpm_original, fpm_replayed
+
+    def log_2_log(
+        self, event_log: str, algorithm: 'AlgoPm4Py'
+    ) -> List[Tuple[str, float]]:
+        """
+        This function takes a log and an algorithm, splits it into multiple sublogs.
+        Next, after running the algorithm on each sublog, the resulting model is being replayed in order to get a new log.
+        Finally, the resulting logs are being compared to the original log.
+
+
+        Parameters:
+        ----------
+        event_log: str
+            Path to the event log.
+        algorithm: AlgoPm4Py
+            Algorithm to be used.
+
+        Returns:
+        -------
+        List of tuples, where each tuple contains the name of the sublog and the conformance value.
+        """
+
+        def split_log_into_sublogs(log):
+            sublogs = []
+            # Initial split to divide the log into two equal parts
+            split_1, split_2 = split_train_test.split(log, train_percentage=0.5)
```

Comment:
From my understanding pm4py's `split_train_test.split()` does actually only shuffle and then randomly samples indices for the train and test log, which can easily be done without the package. If a stratified split is what you want, the scikit-learn `train_test_split()` has an implementation for that, but not easy to define what stratified would be for event log data anyway I guess.